### PR TITLE
rgw_sal_motr: [CORTX-29019] Instrument RGW with ADDB probes

### PR DESCRIPTION
**Pull request by maximalezhin**
Problem: RGW SAL Motr code and incoming RGW requests shall be
instrumented with ADDB probes to reconstruct io-path in scope of
debugging and performance-related work.

Solution: Basic Motr SAL operations (open/write/read/close/delete_obj
and do_idx_op) are instumented with ADDB probes. Handled three types of
operations: S3 Put, S3 Get and S3 Delete multiple objects.

ADDB is Motr sub-system designed for collecting system information
for further analysis. ADDB stands for Analytic and Diagnostic Data-Base.
ADDB probes, added into RGW code, are saved into binary file during RGW
execution and can be converted into human-readable format for
post-execution analysis. Typical ADDB record contains timestamp,
sub-system id (Motr client, CAS, RPC, ...) and passed parameters.

Signed-off-by: Maxim Malezhin <maxim.malezhin@seagate.com>

## Checklist
- Tracker (select at least one)
  - [x] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [x] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [x] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

File: src/rgw/Makefile [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+	@printf "Build RGW ADDB plugin:\n"
+	@printf "\tmake plugin\n"
+	@printf "Remove RGW ADDB plugin:\n"
+	@printf "\tmake clean\n"
+	@printf "Print this message:\n"
+	@printf "\tmake\n"
+	@printf "\tmake help\n"
+	@printf "\n"
+
+plugin: rgw_addb_plugin.so
```
Comment (andriytk) [inline src/rgw/Makefile:22]: ADDB is Motr-backend-specific, right? So, I think, all ADDB-related code should just go to rgw_sal_motr.cc.

Comment (maximalezhin) [reply src/rgw/Makefile:22]: Right, ADDB is Motr-specific, and rgw_sal_motr.cc is the single file in RGW source code tree where ADDB probes were added. There are couple of header files added (rgw_addb.h, rgw_addb_map.h) to keep the code modular. 

Except one thing - unlike Motr, other components that use ADDB probes has to provide so-called ADDB plugin for proper ADDB records parsing. ADDB plugin cannot be placed in rgw_sal_motr.cc file, as it shall be a stand-alone .so binary. Makefile and couple of source files were implemented, but NOT included into the RGW build tree, and proper handling will be tracked in other ticket I about to create for RGW team.

Comment (andriytk) [reply src/rgw/Makefile:22]: > ADDB plugin cannot be placed in rgw_sal_motr.cc file, as it shall be a stand-alone .so binary.

Why?

Comment (maximalezhin) [reply src/rgw/Makefile:22]: Hmm.... I'm afraid my counter-question to your recent question will be "How?" :)
Okay, we might have been having different points in mind. My point is that rgw_addb_plugin.so is supplied to m0addb2dump utility as an input (contains necessary parsers) and m0addb2dump takes probes translation table and makes text dump. 
If we relied on standard RGW build (saying, no Makefile, no addb .so), rgw_sal_motr.cc will be compiled to .o file, but still need to make a shared library from it - and extra step is needed, as m0addb2dump still requires .so file.
Have I understand you correctly that you suggest to make a shared-lib from rgw_sal_motr.cc or what?

Comment (andriytk) [reply src/rgw/Makefile:22]: No, I just wasn't aware that m0addb2dump requires .so. I'm still not sure I understand - why is that so. Why can't it cake radosgw binary as an input?

Comment (maximalezhin) [reply src/rgw/Makefile:22]: Honestly speaking, I haven't tried it and have used a well-known approach developed for S3 server application. 
Is there a strict requirement that no additional files (like plugin .so) shall be added to RGW codebase?

Comment (maximalezhin) [reply src/rgw/Makefile:22]: Conducted an experiment to substitute plugin.so with radosgw binary, and it doesn't work out-of-box. I expect radosgw binary has to be built with specific compiler and linker options (-fPIC, export symbol table, etc.) to behave simultaneously as a binary executable and a shared library. 
m0addb2dump in its turn expects shared library and opens it with dlopen() system call.

Comment (jjxsg) [reply src/rgw/Makefile:22]: > Is there a strict requirement that no additional files (like plugin .so) shall be added to RGW codebase?

@maximalezhin we can add additional files, but we need create a store specific dir and put all extra files there. For motr, please create dir `/src/rgw/store/motr` and move the newly added files there. 
Please check [dbstore](https://github.com/ceph/ceph/tree/master/src/rgw/store/dbstore) for reference.

Comment (andriytk) [reply src/rgw/Makefile:22]: > ...used a well-known approach developed for S3 server application.

Is it required from any motr application to have .so in order to use addb? cc @nikitadanilov

Comment (nikitadanilov) [reply src/rgw/Makefile:22]: m0addb2dump is a utility that parses binary addb stob and produces human-readable addb output. m0addb2dump has a plugin system that allows non-motr addb producers to register parsers for new addb records. So if rgw produces new types of addb records (which it probably does) and we want that standard m0addb2dump would be able to parse there records, rgw should provide a plugin with new parsers.

Comment (maximalezhin) [reply src/rgw/Makefile:22]: Thank you, Nikita, commenting on this. cc: @nikitadanilov 
@andriytk as Nikita said, RGW produces new types of addb records. We don't want to include those types in motr sources (because they are not motr-related), but still want to support them. In this case, plugin, in the form of shared library, is provided.

Comment (andriytk) [reply src/rgw/Makefile:22]: I see...

@maximalezhin, please move this plugin thing to `src/rgw/motr/addb/` directory.

Comment (maximalezhin) [reply src/rgw/Makefile]: @andriytk plugin moved to motr/addb. Would you plz review?

File: src/rgw/rgw_sal_motr.cc [hunk 2]
```diff
@@ -3469,6 +3677,17 @@ int MotrStore::set_buckets_enabled(const DoutPrefixProvider *dpp, vector<rgw_buc
   return 0;
 }
 
+uint64_t MotrStore::get_new_req_id()
+{
+  uint64_t req_id = m0_dummy_id_generate();
```
Comment (andriytk) [inline src/rgw/rgw_sal_motr.cc:3682]: Maybe it would be better to use `ceph::util::generate_random_number<uint64_t>();` ? See also https://github.com/Seagate/cortx-rgw/pull/168.

Comment (maximalezhin) [reply src/rgw/rgw_sal_motr.cc:3682]: Good point. As it's mentioned in #168 , all reqs' ids were set to zero, and both this patch and #168  introduce id assignment. I would agree that using of existing RGW functions better.
The reason why motr function is used here comes from ADDB usage nature and its usage is intentional. `m0_dummy_id_generate()` function guarantees unique id for all callers within motr process. 
ADDB toolkit, PerfLine team developed, uses a tuple of (req_id, process_id) to distinguish each request in motr, regardless what layer request (rgw, client, dix, ioo, rpc, ...) is being processed. If request id source is not the same, then collisions may appear. 
For example, rgw req id is generated with `ceph::util::generate_random_number<uint64_t>()`  and gives `id=N`, and a dix request is generated with `m0_dummy_id_generate()` and it is also id=N. It means that during analysis time, collision of IDs is observed, and it spoils analysis picture.
There is a couple of options to handle it:
1. As RGW req_id may be assigned with an any number, just leave m0_dummy_id_generate(). It works fine and it is the simplest solution.
2. Use `ceph::util::generate_random_number<uint64_t>();` function and implement some more code to map RGW requests to RGW ADDB requests. A bit more code and a bit more complex.
Personally, I prefer to stick with the 1st option.

Comment (maximalezhin) [reply src/rgw/rgw_sal_motr.cc:3682]: What do you think, sir? @andriytk

Comment (andriytk) [reply src/rgw/rgw_sal_motr.cc:3682]: What's the probability of the collision?

**Comment by andriytk** [comment]
@dang, @mattbenjamin, @cbodley, what do you guys think about this patch - will it be acceptable?

File: src/rgw/rgw_sal_motr.cc (hunk 2 above)
Comment (andriytk) [reply src/rgw/rgw_sal_motr.cc:3718]: https://github.com/Seagate/cortx-rgw/pull/168 is merged already, so your code needs to be updated.

File: src/rgw/rgw_sal_motr.cc [hunk 3]
```diff
@@ -4012,6 +4337,46 @@ int MotrStore::init_metadata_cache(const DoutPrefixProvider *dpp,
...
+{
+  for (uint64_t idx = 0; idx < g_rgw_to_addb_idx_func_name_map_size; idx++) {
+    addb_func_to_idx_map[g_rgw_to_addb_idx_func_name_map[idx]] = idx;
+#if defined ADDB_DEBUG
+    ldout(cctx, 0) << "[MM] Registering func: <" << g_rgw_to_addb_idx_func_name_map[idx] << "> with ID: " << idx << dendl;
+#endif
+  }
+}
+
+uint64_t MotrStore::addb_func_to_idx(const std::string& funcname)
```
Comment (andriytk) [inline src/rgw/rgw_sal_motr.cc:4350]: How does it work, and why is this needed?

File: src/rgw/rgw_sal_motr.cc [hunk 4]
```diff
@@ -2435,21 +2629,33 @@ int MotrAtomicWriter::write()
 
     op = nullptr;
     rc = m0_obj_op(obj.mobj, M0_OC_WRITE, &ext, &buf, &attr, 0, 0, &op);
-    if (rc != 0)
+    if (rc != 0) {
       goto err;
+    }
```
[the remark this answers is not in the document]
Comment (maximalezhin) [reply src/rgw/rgw_sal_motr.cc:2634]: Agree

File: src/rgw/rgw_sal_motr.cc (hunk 3 above)
Comment (maximalezhin) [reply src/rgw/rgw_sal_motr.cc:4350]: As ADDB only allows to "log" parameters of uint64_t type, but not strings, lookup table is implemented here.
This function takes function name parameter (__PRETTY_FUNCTION__ macro), finds corresponding index in the  `const char* g_rgw_to_addb_idx_func_name_map[] ` map and returns an index of that function name that will be logged.
ADDB sample added to ADDB stob contains probe_type (i.e. type of request), request id and function idx.
During ADDB stob processing time, m0addb2dump finds probe information from ADDB plugin (rgw_addb_plugin.so) and uses the same map to match function idx with the function name. Function name in its order is added in human-readable way in resulting text dump file.

Comment (nikitadanilov) [reply src/rgw/rgw_sal_motr.cc:4350]: In motr we just log the address of the function (m0_ptr_wrap(&func)) and then in addb2/dump.c use sym() and "-l path-to-shared-library".

File: src/rgw/rgw_sal_motr.cc (hunk 2 above)
Comment (andriytk) [reply src/rgw/rgw_sal_motr.cc:3718]: I'd suggest avoiding "dummy" things. This function is not only used by ADDB, but also by S3 clients. If we use the "dummy" thing - it's very probably that all S3 clients will see relatively the same req ids in the responses - and this will be very inconvenient.

Also, I don't think the issue of collision is so critical for ADDB even when it happens (in a very unlucky case), so it can be ignored.

File: src/rgw/rgw_sal_motr.cc (hunk 3 above)
Comment (andriytk) [reply src/rgw/rgw_sal_motr.cc:4350]: So why cannot we log the address of the function here also, similar like we do in motr?

File: src/rgw/rgw_sal_motr.cc (hunk 2 above)
Comment (maximalezhin) [reply src/rgw/rgw_sal_motr.cc:3718]: What S3 clients are you talking about? Other s3 instances? - still you can distinguish them by FID, and get unique pair of (FID, request_id). 
As for collisions: you might have misunderstand my point - I don't even consider occasional collisions when there is a bullet-proof way of avoiding them.

Still "random" (or whatever you call it as an opposite of "dummy" thing) request id can be assigned, but MotrStore shall implement a map (lookup table) between "random" request_id and "addb" request id. As RGW is multi-threaded application, each thread takes request id from `RGWObjectCTX` parameter and the same request may be executed by different threads, so it's obvious why lookup is need. But how map is cleaned up? Is there a function call or method to indicate that a request processing is finished?

**Comment by mattbenjamin** [comment]
@maximalezhin @andriytk what is ADDB/an ADDB plugin?  it is not a framework I'm familiar with.  We have work in progress to add jaeger tracing to high level RGW workflows, and I don't think we'd preempt that.  I don't think anyone would object to motr including trace logic that lives in its motr codebase, though.

File: src/rgw/rgw_sal_motr.cc (hunk 3 above)
Comment (maximalezhin) [reply src/rgw/rgw_sal_motr.cc:4350]: Two reasons:
1. I had no clue about m0_ptr_wrap() and function name resolving. Thanks @nikitadanilov for pointing the right direction. S3server code implements the same function <-> idx map as implemented here, and just reused the code.
2. @andriytk Does RGW build system generate shared (or static) library? If yes, what is it? Can libbfd() deal with C++ symbols? I suggest to take function address-name resolution as another ticket (and PR)

**Comment by maximalezhin** [comment]
@mattbenjamin ADDB is Motr sub-system designed for collecting system information for further analysis. ADDB stands for Analytic and Diagnostic Data-Base. ADDB probes, added into RGW code, are saved into binary file during RGW execution and can be converted into human-readable format for post-execution analysis. Typical ADDB record contains timestamp,
sub-system id (Motr client, CAS, RPC, ...) and passed parameters.

**Comment by andriytk** [comment]
> [rgw_sal_motr: [CORTX-29019] Address review comments](https://github.com/Seagate/cortx-rgw/pull/142/commits/6bc41678570444904a7de910fa38a03783535944)

@maximalezhin you can squash this into the main patch.

**Review by andriytk — APPROVED** [review]

**Comment by maximalezhin** [comment]
Squashed into a single commit.

File: src/rgw/rgw_sal_motr.cc [hunk 5]
```diff
@@ -3656,6 +3857,18 @@ int MotrStore::set_buckets_enabled(const DoutPrefixProvider *dpp, vector<rgw_buc
   return 0;
 }
 
+uint64_t MotrStore::get_new_req_id()
+{
+  uint64_t req_id = ceph::util::generate_random_number<uint64_t>();
```
Comment (nikitadanilov) [inline src/rgw/rgw_sal_motr.cc:3862]: This introduces a collision possibility for no good reason. Can a sequential atomic be used instead?

**Review by nikitadanilov — CHANGES_REQUESTED** [review]
Binary is in commit: src/rgw/motr/addb/rgw_addb_plugin.so ?

File: src/rgw/rgw_sal_motr.cc (hunk 5 above)
Comment (andriytk) [reply src/rgw/rgw_sal_motr.cc:3862]: @nikitadanilov, this id is used by S3 clients in responses. With sequential atomic, it's very probable that S3 clients will get the replies with the same ids from different rgw instances.

File: src/rgw/motr/addb/rgw_addb.h [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+ */
+
+#pragma once
+
+#ifndef __RGW_ADDB_H__
+#define __RGW_ADDB_H__
+
+#include <addb2/addb2_internal.h>
+
+#define ADDB(_addb_id, ...)                                      \
```
[the remark this answers is not in the document]
Comment (maximalezhin) [reply src/rgw/motr/addb/rgw_addb.h:22]: This option has been tried, and g++ throws an error while expanding of M0_ADDB2_ADD() macro:
`/usr/include/motr/addb2/addb2.h:254:26: error: taking address of temporary array
        (const uint64_t[]){ __VA_ARGS__ })
`
Agree that proposed solution does apprx. the same, but declares an array of parameters on stack.

**Comment by maximalezhin** [comment]
> Binary is in commit: src/rgw/motr/addb/rgw_addb_plugin.so ?

Yes, it is a workaround unless plugin is built during RPM creation. Corresponding ticket is created, and binary will be deleted when solution is ready. Does it sound good to you?

File: src/rgw/rgw_sal_motr.cc [hunk 7]
```diff
@@ -54,6 +57,41 @@ namespace rgw::sal {
...
+  uint64_t req_id;
+  bool motr_thread = false;
+public:
+  MotrADDBLogger() {
+    struct m0_thread_tls *tls = m0_thread_tls();
+
+    req_id = (uint64_t)-1;
+    if (tls->tls_addb2_mach == NULL)
+    {
+      m0_addb2_global_thread_enter();
```
[the remark this answers is not in the document]
Comment (maximalezhin) [reply src/rgw/rgw_sal_motr.cc:71]: I'll split my answer into to parts:
1. Why tls is not checked for NULL:
This is because of `m0_thread_tls()` function call - it returns pointer to thread-local-store instance of `struct m0_thread_tls`, and even if tls is NULL, it will allocate and return.
2. Why tls != NULL and tls->tls_addb2_mach == NULL:
Firstly it should be said that RGW is multithreaded application, and only one thread is initialized as Motr thread ( with tls, addb2_mach and all internal stuff). Other threads have to be initialized, or "entered" as a motr global thread. When a thread creates an instance of `class MotrADDBLogger` it doesn't have any tls (and all related variables) initialized, hence any ADDB call will be just ignored - as tls_addb2_mach is not allocated yet. Code above allows thread to operate as a motr thread, and `m0_addb2_global_thread_enter()` allocates addb machine for following logging.

File: src/rgw/rgw_sal_motr.cc (hunk 5 above)
Comment (nikitadanilov) [reply src/rgw/rgw_sal_motr.cc:3862]: But collision is still possible with random identifiers. Does a situation when a client gets the same identifiers from different instances lead to a problem?

**Comment by nikitadanilov** [comment]
> > Binary is in commit: src/rgw/motr/addb/rgw_addb_plugin.so ?
> 
> Yes, it is a workaround unless plugin is built during RPM creation. Corresponding ticket is created, and binary will be deleted when solution is ready. Does it sound good to you?

Yes, this is fine. Please reference the ticket in the PR.

File: src/rgw/rgw_sal_motr.cc [hunk 8]
```diff
@@ -4136,6 +4432,46 @@ int MotrStore::init_metadata_cache(const DoutPrefixProvider *dpp,
...
+  for (const auto& v : addb_func_to_idx_map) {
+    ldout(cctx, 0) << "ADDB Debug: func:<"  << v.first << ">, idx:<" << v.second << ">" << dendl;
+  }
+#endif
+
+  if (addb_func_to_idx_map.count(func) == 0) {
+    ldout(cctx, 0) << "unknown ADDB function: " << func << dendl;
+    idx = 0;
+  } else {
+    idx = addb_func_to_idx_map[func];
```
[the remark this answers is not in the document]
Comment (maximalezhin) [reply src/rgw/rgw_sal_motr.cc:4465]: It was an opposite intention - to avoid static list (e.g., enum, array) of function and give more flexibility. By using PRETTY_FUNCTION it's impelemented partially - we still maintain array for idx->function_name_str conversion.
Having enum requires developer to pass exact corresponding value for a function, and it's a place of error.
Do you have alternative view on it? I assumed PRETTY_FUNCTION macro and map lookup doesn't affect performance, but what are your ideas? 
Let's schedule a call and discuss.

**Comment by maximalezhin** [comment]
@nikitadanilov Nikita would you please review the PR?

File: src/rgw/motr/addb/rgw_addb_map.c [hunk 9]
```diff
@@ -0,0 +1,43 @@
...
+  "rgw::sal::MotrStore::create_motr_idx_by_name"
+};
+
+const uint64_t g_rgw_to_addb_idx_func_name_map_size = 13;
+
+const char* addb_idx_to_rgw_state(uint64_t map_idx) {
+  assert(map_idx < g_rgw_to_addb_idx_func_name_map_size);
+
+  return g_rgw_to_addb_idx_func_name_map[map_idx];
+}
```
Comment (nikitadanilov) [inline src/rgw/motr/addb/rgw_addb_map.c:42]: Is this still needed?

Comment (maximalezhin) [reply src/rgw/motr/addb/rgw_addb_map.c]: Missed to remove. Removed within new commit.

**Review by nikitadanilov — APPROVED** [review]

**Review by jjxsg — APPROVED** [review]

**Comment by cdeshmukh** [comment]
@maximalezhin needs rebase and conflict resolution

File: src/rgw/rgw_sal_motr.h [hunk 10]
```diff
@@ -22,6 +22,8 @@ extern "C" {
 #pragma clang diagnostic ignored "-Wextern-c-compat"
 #include "motr/config.h"
 #include "motr/client.h"
+#include "lib/thread.h"
```
Comment (andriytk) [inline src/rgw/rgw_sal_motr.h:25]: Shouldn't it be `"motr/lib/thread.h"` ?

File: src/rgw/rgw_sal_motr.cc [hunk 11]
```diff
@@ -1172,10 +1229,6 @@ void MotrStore::finalize(void)
   m0_client_fini(this->instance, true);
 }
 
-uint64_t MotrStore::get_new_req_id() {
```
Comment (andriytk) [inline src/rgw/rgw_sal_motr.cc:1175]: Why was it moved?

**Comment by andriytk** [comment]
Is there some documentation/readme on how to configure and use it?

File: src/rgw/rgw_sal_motr.cc (hunk 11 above)
Comment (maximalezhin) [reply src/rgw/rgw_sal_motr.cc:1175]: It's actually there, line no. 4395. 
Cause or removing this and adding that- merging artifact.

Comment (maximalezhin) [reply src/rgw/rgw_sal_motr.cc:1175]: Yes, sir. Done.

File: src/rgw/rgw_sal_motr.h (hunk 10 above)
Comment (maximalezhin) [reply src/rgw/rgw_sal_motr.h]: No, just `lib/thread`. It's needed for adoptation of non-motr thread for ADDB logging.

**Comment by maximalezhin** [comment]
ADDB description is added into commit msg.
ADDB subsystem and its usage can be found in Confluence area: https://seagate-systems.atlassian.net/wiki/spaces/PRIVATECOR/pages/891978152/Motr+ADDB+training

**Comment by maximalezhin** [comment]
Requesting @nikitadanilov @jjxsg @sachinpunadikar @bkirunge7 to review this patch.
Originally it was already approved (got +3), but as plugin code has been removed (major change though, and merged into cortx-rgw-integration repo), re-requesting another +1 and merge into main.

**Review by jjxsg — APPROVED** [review]

File: src/rgw/rgw_sal_motr.cc [hunk 12]
```diff
@@ -1627,8 +1695,11 @@ int MotrObject::MotrReadOp::read(int64_t off, int64_t end, bufferlist& bl, optio
 // a block of data each time and call 'cb' for post-processing.
 int MotrObject::MotrReadOp::iterate(const DoutPrefixProvider* dpp, int64_t off, int64_t end, RGWGetDataCB* cb, optional_yield y)
 {
+
```
Comment (andriytk) [inline src/rgw/rgw_sal_motr.cc:1698]: Nitpick: drop this empty line.

**Comment by andriytk** [comment]
> ADDB subsystem and its usage can be found in Confluence area: https://seagate-systems.atlassian.net/wiki/spaces/PRIVATECOR/pages/891978152/Motr+ADDB+training

Can we also add a README.md file at `src/rgw/motr/addb/` with a quickstart info on how to configure and use it?

**Comment by maximalezhin** [comment]
> 

I suggest to take README.md to take as a separate PR. Currently no tuning and configurations are expected from this patch.
RGW ADDB is supposed to work as a part of PerfLine utility run cycles, and all information to be collected automatically w/o user involvement.

File: src/rgw/rgw_sal_motr.cc (hunk 12 above)
Comment (maximalezhin) [reply src/rgw/rgw_sal_motr.cc]: Done

**Review by nikitadanilov — APPROVED** [review]

**Review by andriytk — APPROVED** [review]
Thank you!

File: src/rgw/rgw_sal_motr.cc [hunk 13]
```diff
@@ -2474,24 +2660,37 @@ int MotrObject::read_mobj(const DoutPrefixProvider* dpp, int64_t off, int64_t en
...
+	   RGW_ADDB_FUNC_READ_MOBJ,
+	   RGW_ADDB_PHASE_ERROR);
       goto out;
     }
     // Call `cb` to process returned data.
     ldpp_dout(dpp, 20) << __func__  << " call cb to process data" << dendl;
     cb->handle_data(bl, bloff, actual);
     bloff = 0;
+
+    addb_logger.set_id(req_id);
```
Comment (andriytk) [inline src/rgw/rgw_sal_motr.cc:2687]: Why is this needed?

Comment (maximalezhin) [reply src/rgw/rgw_sal_motr.cc:2687]: The reason why this at first glance not necessary call is needed relates to the nature of multithreading in RGW.
You might have seen that `addb_logger` object resides in thread local storage. It is needed because it allows to associate each thread with particular unique ID that will be logged into ADDB stob. It also allows us to avoid modification of RGW sources and map a RGW request ID to subsequent motr client request IDs - to plot request sequential diagrams, histograms, etc. 
So, ideally, once request ID is assigned to thread local storage (i.e., addb_logger), it might be used from any place of the code. But it is not true for RGW threading model. RGW is based on asio threads. and its engine may (and actually does) change thread working on a particular request. Assume req0 is processed by thread0, and and thread0 is assigned with req0 ID. Next, `cb->handle_data()` is executed, and req0 is now executed by thread1, while thread0 is returned back to thread pool. And it means we have to re-assign req ID to thread1 so it can be dumped correctly.
